### PR TITLE
RATEST-117:EditPatientRelationshipPage

### PR DESCRIPTION
**Pull request by sherrif10**
https://issues.openmrs.org/browse/RATEST-117  cc @k-joseph

### CODE REVIEW

File: ui-tests/src/test/java/org/openmrs/reference/EditPatientRelationshipTest.java
Comment (kdaud) [inline ui-tests/src/test/java/org/openmrs/reference/EditPatientRelationshipTest.java:20]: Refactor this function out, its not required within the resource

File: ui-tests/src/test/java/org/openmrs/reference/EditPatientRelationshipTest.java
Comment (kdaud) [inline ui-tests/src/test/java/org/openmrs/reference/EditPatientRelationshipTest.java:28]: You do not need this annotation here

File: ui-tests/src/test/java/org/openmrs/reference/EditPatientRelationshipTest.java
Comment (kdaud) [inline ui-tests/src/test/java/org/openmrs/reference/EditPatientRelationshipTest.java:29]: same here

File: ui-tests/src/test/java/org/openmrs/reference/EditPatientRelationshipTest.java
Comment (kdaud) [inline ui-tests/src/test/java/org/openmrs/reference/EditPatientRelationshipTest.java:46]: same here

File: ui-tests/src/test/java/org/openmrs/reference/EditPatientRelationshipTest.java
Comment (kaweesi) [inline ui-tests/src/test/java/org/openmrs/reference/EditPatientRelationshipTest.java:11]: We don't need this test, we rather want to have this written in bdd steps

**Review by kaweesi — CHANGES_REQUESTED** [review]
this change should only include pages and not a test

File: ui-tests/src/test/java/org/openmrs/reference/EditPatientRelationshipTest.java
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/EditPatientRelationshipTest.java:11]: Ohh didnt knew this thanks let me remove this class

**Comment by sherrif10** [comment]
Thanks so much @k-joseph  , i have done as you had suggested

File: ui-tests/src/test/java/org/openmrs/reference/page/EditPatientRelationshipPage.java
Comment (sherrif10) [inline ui-tests/src/test/java/org/openmrs/reference/page/EditPatientRelationshipPage.java:23]: Hello @kaweesi , I need  your help with in this line, i want to tell the driver to be able to loop through different relationshipstype not only restricting to only doctor attribute, inotherwords the driver can always look through different types of relationship, how can we implement this thanks

**Comment by sherrif10** [comment]
cc @kaweesi  Kindly review   thanks

File: ui-tests/src/test/java/org/openmrs/reference/page/EditPatientRelationshipPage.java
Comment (kaweesi) [reply ui-tests/src/test/java/org/openmrs/reference/page/EditPatientRelationshipPage.java:23]: you can pass a relationship type element into the method which you can pass into the feature file.

Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/EditPatientRelationshipPage.java:23]: With in the feature file , i passed doctor as relationship type ,  are you meaning something on this line
 relationshipType.selectByVisibleText("Doctor");

File: ui-tests/src/test/java/org/openmrs/reference/page/EditPatientRelationshipPage.java
Comment (kaweesi) [inline ui-tests/src/test/java/org/openmrs/reference/page/EditPatientRelationshipPage.java:22]: next thing is select a person, ensure your test creates that person in before and removes him and the relationship in after

**Review by kaweesi — CHANGES_REQUESTED** [review]
good starting point @sherrif10 please check my comment to get this test completely useful

File: ui-tests/src/test/java/org/openmrs/reference/page/EditPatientRelationshipPage.java
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/EditPatientRelationshipPage.java:22]: Let me ask , is it neccessary to create a person when we can search for existing person

**Comment by sherrif10** [comment]
Have resolved by adding an element that takes person name  thanks kindly review

**Comment by sherrif10** [comment]
cc @kaweesi  KIndly waiting for on your final review on this thanks

File: ui-tests/src/test/java/org/openmrs/reference/page/ClinicianFacingPatientDashboardPage.java
Comment (kaweesi) [inline ui-tests/src/test/java/org/openmrs/reference/page/ClinicianFacingPatientDashboardPage.java:84]: now that the id contains ".", use an attribute selector rather

File: ui-tests/src/test/java/org/openmrs/reference/page/ClinicianFacingPatientDashboardPage.java
Comment (kaweesi) [inline ui-tests/src/test/java/org/openmrs/reference/page/ClinicianFacingPatientDashboardPage.java:84]: do you really mean https://css-tricks.com/whats-the-difference/#:~:text=A%20space%20in%20a%20CSS,because%20it%20has%20no%20space.?

an attribute selector that i suggested looks like By.cssSelector("a[id='application.registrationapp.summary.editPatientLink']")

Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/ClinicianFacingPatientDashboardPage.java]: Resolved thanks.

Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/ClinicianFacingPatientDashboardPage.java]: Hello @k-joseph  Have replied to all your comments looking forward to hear from you

File: ui-tests/src/test/java/org/openmrs/reference/page/EditPatientRelationshipPage.java
Comment (k-joseph) [reply ui-tests/src/test/java/org/openmrs/reference/page/EditPatientRelationshipPage.java:24]: it's ok to select an existing patient

File: ui-tests/src/test/java/org/openmrs/reference/page/ClinicianFacingPatientDashboardPage.java
Comment (k-joseph) [inline ui-tests/src/test/java/org/openmrs/reference/page/ClinicianFacingPatientDashboardPage.java:83]: you can get rid of this trailing space

File: ui-tests/src/test/java/org/openmrs/reference/page/EditPatientRelationshipPage.java
Comment (k-joseph) [inline ui-tests/src/test/java/org/openmrs/reference/page/EditPatientRelationshipPage.java:25]: add a space between `NAME,` and `name`

File: ui-tests/src/test/java/org/openmrs/reference/page/EditPatientRelationshipPage.java
Comment (k-joseph) [inline ui-tests/src/test/java/org/openmrs/reference/page/EditPatientRelationshipPage.java:12]: this would be selecting all buttons and all inputs, please correct this to only set the person's name

File: ui-tests/src/test/java/org/openmrs/reference/page/EditPatientRelationshipPage.java
Comment (k-joseph) [inline ui-tests/src/test/java/org/openmrs/reference/page/EditPatientRelationshipPage.java:26]: before this, where is the step that selects `John` from the popup with results?

Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/EditPatientRelationshipPage.java:26]: setTextToFieldNoEnter(NAME,name); allows the input of the string "john"

File: ui-tests/src/test/java/org/openmrs/reference/page/EditPatientRelationshipPage.java
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/EditPatientRelationshipPage.java:24]: This cssSelector element allows input of person name

**Comment by sherrif10** [comment]
Hello @k-joseph  @kdaud  is this yet worth it

File: ui-tests/src/test/java/org/openmrs/reference/page/ClinicianFacingPatientDashboardPage.java
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/ClinicianFacingPatientDashboardPage.java]: Resolved thanks cc @k-joseph

**Comment by sherrif10** [comment]
Hello @ibacher @dkayiwa  Kindly review if you find time , This Looks Good to me

**Comment by dkayiwa** [comment]
@sherrif10 my suggestion is that, before you call on us for review, can you first ensure that @kdaud has reviewed and approved? The reverse also applies where @sherrif10 first reviews and approves @kdaud's pull requests before calling up on us. Deal?

**Comment by sherrif10** [comment]
Sure Deal thanks @dkayiwa

File: ui-tests/src/test/java/org/openmrs/reference/page/EditPatientRelationshipPage.java
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/EditPatientRelationshipPage.java]: setTextToFieldNoEnter(NAME,name); allows the input of the string "john"

**Comment by sherrif10** [comment]
cc @kdaud  do you mind looking into this, before we ping @dkayiwa  @ibacher for final review

File: ui-tests/src/test/java/org/openmrs/reference/page/ClinicianFacingPatientDashboardPage.java
Comment (kdaud) [inline ui-tests/src/test/java/org/openmrs/reference/page/ClinicianFacingPatientDashboardPage.java:78]: Is the change necessary at code line `LN-78` ?

File: ui-tests/src/test/java/org/openmrs/reference/page/EditPatientRelationshipPage.java
Comment (kdaud) [inline ui-tests/src/test/java/org/openmrs/reference/page/EditPatientRelationshipPage.java:12]: If you get a better element reference for `PERSON_NAME` would be a good deal. One challenge with locating a page element through `nth-child(x)` value is that things get broken when UI changes comes in and element can not be located any more via `x-` value!

File: ui-tests/src/test/java/org/openmrs/reference/page/RegistrationSummaryPage.java
Comment (kdaud) [inline ui-tests/src/test/java/org/openmrs/reference/page/RegistrationSummaryPage.java:11]: Remove one of the redundant line code `LN-10` or `LN-11` for quality assurance.

**Comment by kdaud** [comment]
Thanks @sherrif10 for the work!
Catch my reviews above!

File: ui-tests/src/test/java/org/openmrs/reference/page/ClinicianFacingPatientDashboardPage.java
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/ClinicianFacingPatientDashboardPage.java:78]: There is no change here thanks

File: ui-tests/src/test/java/org/openmrs/reference/page/RegistrationSummaryPage.java
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/RegistrationSummaryPage.java]: Resolved thanks

**Comment by sherrif10** [comment]
Hello @kdaud  any more reviews or recommendations

File: ui-tests/src/test/java/org/openmrs/reference/page/EditPatientRelationshipPage.java
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/EditPatientRelationshipPage.java:12]: i would go with this cssSelector element , Alternatively using xpath

**Comment by sherrif10** [comment]
cc @dkayiwa  @ibacher

**Comment by dkayiwa** [comment]
@sherrif10 before pinging us, have you checked asked for @kdaud's approval?

**Comment by sherrif10** [comment]
cc @kdaud

File: ui-tests/src/test/java/org/openmrs/reference/page/ClinicianFacingPatientDashboardPage.java
Comment (kdaud) [inline ui-tests/src/test/java/org/openmrs/reference/page/ClinicianFacingPatientDashboardPage.java:75]: Should we not get a better element reference since `linkText` is not reliable? How about using `cssSelector`, or id or `xPath`

File: ui-tests/src/test/java/org/openmrs/reference/page/EditPatientRelationshipPage.java
Comment (kdaud) [reply ui-tests/src/test/java/org/openmrs/reference/page/EditPatientRelationshipPage.java:12]: @sherrif10 did you see the concern on current `cssSelector` value ? I think we need to put code sustainability in mind to reduce the maintenance burden. If we would go with the current set value this means time will come when that value will break the test due to the reason I gave above this is another pain looking through the test locating where things are breaking.

**Comment by kdaud** [comment]
Great effort @sherrif10!

 I have left one or two reviews for you.

File: ui-tests/src/test/java/org/openmrs/reference/page/ClinicianFacingPatientDashboardPage.java
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/ClinicianFacingPatientDashboardPage.java:75]: Actually this should be removed we noloner support eye report

Comment (kdaud) [reply ui-tests/src/test/java/org/openmrs/reference/page/ClinicianFacingPatientDashboardPage.java:75]: > Eye report is no longer supported as you have been seeing different conversations we normally had

But git is tracking this as your change. If its not true then I recommend you do some settings in your IDE for git only to reflect your changes!

Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/ClinicianFacingPatientDashboardPage.java:75]: i wouldnt to touch on this file, i would like to leave as it is since  eye ReportTest class is still in the master

Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/ClinicianFacingPatientDashboardPage.java:75]: Unless we clearly remove it which can be worked on a different PR.

Comment (kdaud) [reply ui-tests/src/test/java/org/openmrs/reference/page/ClinicianFacingPatientDashboardPage.java:75]: > Unless we clearly remove it which can be worked on a different PR.

Are you sure that its not used by some instances ?. If no we can still use a different element look up to ensure code sustainability otherwise `linkText` is not smart enough

Comment (kdaud) [reply ui-tests/src/test/java/org/openmrs/reference/page/ClinicianFacingPatientDashboardPage.java:75]: > i wouldnt to touch on this file, i would like to leave as it is since eye ReportTest class is still in the master

Do you mind sharing a link to this file ?  

BTW  did you see my recommendation [here](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/382#discussion_r682728044) !

**Pull request closed without merging**
